### PR TITLE
Fix clang-tidy warnings in HIP

### PR DESCRIPTION
**Pull request by Rombur**
This PR fixes the clang-tidy warnings in HIP. I've added a new configuration for Frontier and moved the yml file to `.gitlab`. I need to use `amdclang` to use `clang-tidy`. It doesn't work with `hipcc`.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 1]
```diff
@@ -330,7 +330,7 @@ class HIPTeamMember {
                 const size_t scratch_level_1_size, const int arg_league_rank,
                 const int arg_league_size)
       : m_team_reduce(shared),
-        m_team_shared(((char*)shared) + shared_begin, shared_size,
+        m_team_shared(reinterpret_cast<char*>(shared) + shared_begin, shared_size,
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Team.hpp:333]:
```suggestion
        m_team_shared(static_cast<char*>(shared) + shared_begin, shared_size,
```

**Review by dalg24 — COMMENTED** [review]
Thoughts on also adding to one of the CI builds that reports back a "check", that is one of the Jenkins ORNL CI builds or the MI300A build on Spack GitLab

File: core/src/HIP/Kokkos_HIP_Team.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Team.hpp:333]: agreed

File: core/src/HIP/Kokkos_HIP.cpp [hunk 2]
```diff
@@ -96,7 +96,7 @@ void HIP::impl_initialize(InitializationSettings const& settings) {
   // Allocate a staging buffer for constant mem in pinned host memory
   // and an event to avoid overwriting driver for previous kernel launches
   KOKKOS_IMPL_HIP_SAFE_CALL(
-      hipHostMalloc((void**)&Impl::HIPInternal::constantMemHostStaging,
+      hipHostMalloc(reinterpret_cast<void**>(&Impl::HIPInternal::constantMemHostStaging),
                     Impl::HIPTraits::ConstantMemoryUsage));
```
Comment (tpadioleau) [inline core/src/HIP/Kokkos_HIP.cpp:100]: Can I invoke memory safety concerns (https://www.c-faq.com/ptrs/genericpp.html) to replace this `reinterpret_cast` with something like
```cpp
  void* void_ptr = nullptr;
  KOKKOS_IMPL_HIP_SAFE_CALL(hipHostMalloc(&void_ptr, Impl::HIPTraits::ConstantMemoryUsage));
  Impl::HIPInternal::constantMemHostStaging = void_ptr;
```

**Review by tpadioleau — COMMENTED** [review]
Cool thanks

**Comment by Rombur** [comment]
I've added a build in jenkins that uses clang-tidy. The ROCm 6.2 image does not contain clang-tidy, so I picked ROCm 6.0

**Review by masterleinad — APPROVED** [review]
I'm fine with addressing the `reinterpret_cast<void**>` issues elsewhere for all occurrences.

File: .jenkins [hunk 3]
```diff
@@ -340,6 +340,41 @@ pipeline {
                         }
                     }
                 }
+                stage('HIP-ROCm-6.0') {
```
Comment (dalg24) [inline .jenkins:343]: Do you anything else in the name of the build to indicate that this one, in contrast to the other that set hipcc as the cxx compiler, uses amdclang directly?

File: core/src/HIP/Kokkos_HIP.cpp (hunk 2 above)
Comment (dalg24) [reply core/src/HIP/Kokkos_HIP.cpp]: I agree that this is something we should consider fixing.
Considering that I was able to find a few more instances of this pattern with a quick search, I am fine with us doing it as a follow up.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**